### PR TITLE
Implement support for .NET Standard 2.0

### PR DESCRIPTION
**Pull request by Nihlus**
Many libraries are still using .NET Standard 2.0, and it would be quite useful for them to be able to consume this library.

This PR implements support for .NET Standard 2.0 in the library, utilizing some minor compile-time preprocessing to ensure that .NET Core targets don't lose any performance because of these changes.

If you merge this pull request, I'd also appreciate it if you tagged it as "hacktoberfest-accepted", so that it counts towards a t-shirt :P

### CODE REVIEW

File: src/Qmmands/Builders/CommandBuilder.cs [hunk 1]
```diff
@@ -97,7 +97,14 @@ internal CommandBuilder(ModuleBuilder module, object callback)
             Module = module;
             Callback = callback;
             Cooldowns = new List<Cooldown>();
+
+            #if NETSTANDARD2_0
+            Aliases = new HashSet<string>();
```
Comment (Emzi0767) [inline src/Qmmands/Builders/CommandBuilder.cs:102]: This change is not a net positive. It causes the set to be created with whatever defaults are there, which might cause growing and reallocations. Standard 2.0 should very much support initial capacity argument.

File: src/Qmmands/CommandService.cs [hunk 2]
```diff
@@ -124,7 +124,11 @@ public CommandService(CommandServiceConfiguration configuration)
                 ? configuration.NullableNouns.ToImmutableArray()
                 : CommandUtilities.DefaultNullableNouns;
 
+            #if NETSTANDARD2_0
+            StringComparer = StringComparerExtensions.FromComparison(StringComparison);
```
Comment (Emzi0767) [inline src/Qmmands/CommandService.cs:128]: What is the point of this?

File: src/Qmmands/Extensions/StringComparerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+using System;
+
+namespace Qmmands
+{
+    /// <summary>
+    ///     Defines extensions for the <see cref="StringComparer"/> class.
+    /// </summary>
+    public class StringComparerExtensions
```
Comment (Emzi0767) [inline src/Qmmands/Extensions/StringComparerExtensions.cs:8]: Same as previous comment.

**Review by Emzi0767 — CHANGES_REQUESTED** [review]
There's little point in targeting standard 2.0. Existing core targets cover all the intended use cases. You really should switch away from .NETFX.

File: src/Qmmands/Builders/CommandBuilder.cs (hunk 1 above)
Comment (Nihlus) [reply src/Qmmands/Builders/CommandBuilder.cs:102]: Standard 2.0 does not support that argument, unfortunately.

File: src/Qmmands/CommandService.cs (hunk 2 above)
Comment (Nihlus) [reply src/Qmmands/CommandService.cs:128]: .NET Standard 2.0 does not define the FromComparison method.

File: src/Qmmands/Extensions/StringComparerExtensions.cs (hunk 3 above)
Comment (Nihlus) [reply src/Qmmands/Extensions/StringComparerExtensions.cs:8]: See above.

**Comment by Nihlus** [comment]
I don't use NETFX, but .NET Standard 2.0 is a valid target for many existing libraries and plugin modules for various bots and command services. The changes are minor, and those who target a .netcoreappx.x target will not notice any difference. It's also a benefit to target at least one standard, since .netcoreapp is - by default - a runnable executable target, and netstandard is explicitly for libraries.

An alternative would be targeting netstandard2.1, which does not require any source changes at all. This does preclude NETFX, of course, but still retains support for Mono, Xamarin, Unity, and UWP (which are not supported at all when there's only netcoreapp targets).

**Comment by Emzi0767** [comment]
.NET Standard 2.1 is literally just Core 3.0. There's nothing else in that target. If you do not use FX, then realistically the only target for scenarios outlined is Core 3.x. Which is already targeted.

**Comment by Nihlus** [comment]
That's not strictly true - .NET Standard 2.1 is supported by many other frameworks and runtimes other than .NET Core 3.x, as I mentioned above. Those targets can't consume the library in its current targeting setup - it's limited to .NET Core only, and nothing else. This will also matter in a few months when .NET 5 has been released, wherein it won't be able to use the library either unless it either targets net5 or a netstandard target.

**Comment by Emzi0767** [comment]
Go check the target matrix on .NET's website. 

Furthermore, targeting smart fridges is pointless. UWP is already .NET Core, and I can't really think of any other realistic targets, and even UWP isn't one.

Plus .NET 5 is an evolution of Core. It will be able to consume it.

**Comment by Nihlus** [comment]
Fair enough, in regards to .NET 5 - I hadn't seen that it would support direct references of netcoreapp targets. Again, though, targeting netstandard2.1 requires no source changes and allows other runtimes other than .NET Core to consume the library - there's no downside whatsoever, if the changes in this pull request to allow netstandard2.0 are unwanted.

The [support matrix](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) for .NET Standard paints a pretty clear picture, in my opinion - .NET Core is a single target framework out of eight listed, and that's not counting any future implementations that don't exist yet. Targeting netstandard is like using a C++ standard or a standard screw size - it not only ensures compatibility with current technologies, but future ones as well.

Even Microsoft themselves recommend targeting .NET Standard for libraries, and netstandard2.0 in particular (see [this](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting) article). And, yes, while .NET 5 will be the de facto evolution of both .NET Core and .NET Standard, we're not there yet, and .NET Standard will be relevant for a long time - see [this](https://devblogs.microsoft.com/dotnet/the-future-of-net-standard/) article.

**Comment by Emzi0767** [comment]
https://dotnet.microsoft.com/platform/dotnet-standard

Mono is a broken target so there's a good chance it won't run at all even if it says supported. UWP is a Core 2.x target, and along with Xamarin is in the smart fridge category. 

Future targets will target .NET 5, which is about to hit GA, with the possible exception of embedded or specialized stuff that most likely won't target Standard at all.

Standard is going the way of PCL, it's getting replaced by a more universal target.

**Comment by Nihlus** [comment]
I disagree with that analysis, but even if I didn't, why not target netstandard2.1 in the interim? There's no good reason not to - many libraries and applications will remain on the old targets for a significant amount of time, and libraries that purely target netstandard can't consume this library as it is right now.

Again, netstandard2.1 is a single-line change that requires no other source changes, but opens up compatibility for a much, much wider audience than you have today. You can't possibly know every potential use case for your library, so why close those doors without a good reason?

I should clarify; I'm not asking you to drop netcore targets, just add netstandard as an additional option for those of us who have purely netstandard libraries.

**Comment by Emzi0767** [comment]
This is why

Core 3.x is already covered. So that's why.

**Comment by Perksey** [comment]
The advise from Microsoft on .NET Standard is to target the lowest version you are capable of supporting. For many people, this is 2.0 because 1.x is severely lacking for a lot of workloads, just as .NET Core 1.x was back in the days.

It's also worth noting that UWP is not .NET Core. During debugging, .NET Core is used to emulate UWP yes￼, however when compiled .NET Native does not explicitly support .NET Core stuff (it since day one has used a subset ambiguously named .NET Core 5.0, but this was made before the .NET Core we know today was made)

.NET Framework is still in active support and, even though .NET Core is a recommended switch, many library authors would like to support all targets (for example our policy is use .NET Standard 2.0 until .NET 6 comes out with mobile support).

.NET Standard 2.0 is the de facto target for libraries and I'm sure many of Qmmands' potential users will have wanted to use it in this context.

**Comment by Nihlus** [comment]
You're missing the point - a library which, today, targets netstandard and not netcore is unable to also depend on this library. The only way for it to target this library is to add a netcoreapp target to itself, which also means that the netstandard target is strictly incompatible, and must be removed.

In contrast, if this library adds a netstandard2.1 target, no changes are required, and it can be used by a wider audience. 

Core 3.x is covered, yes, but libraries don't tend to target Core. They tend to target netstandard, and no framework-specific targets. Those libraries cannot depend on this library in the current setup without also dropping their netstandard targets.

**Comment by Perksey** [comment]
Cc @GreemDev may be worth making a fork if this change doesn't go in, as it's clearly something that could be in demand.

**Comment by Emzi0767** [comment]
> Core 3.x is covered, yes, but libraries don't tend to target Core.

This library targets applications, not libraries. Applications target Core, not Standard.

> It's also worth noting that UWP is not .NET Core. During debugging, .NET Core is used to emulate UWP yes￼, however when compiled .NET Native does not explicitly support .NET Core stuff (it since day one has used a subset ambiguously named .NET Core 5.0, but this was made before the .NET Core we know today was made)

dnx50, which is that which you're referring to, is Core 1.0, and wouldn't be supported regardless of whether this change is approved or not.

> .NET Standard 2.0 is the de facto target for libraries and I'm sure many of Qmmands' potential users will have wanted to use it in this context.

So far I only see 3. You, Perksey, and Greem. And it's because of bad project structure.

**Comment by Perksey** [comment]
Hang on, just realized that Emzi0767 isn't a maintainer, so will wait for @Quahu to respond before doing anything drastic like forking.

**Comment by Quahu** [comment]
It'd be foolish to disagree with what Emzi has said so far, thus consider it my words. Most of it is facts, not opinions. I have and might again consider Standard 2.1, but Standard 2.0 is a complete waste.

**Comment by Nihlus** [comment]
Wrapper libraries are a thing, as are modular codebases. Not everything that wants to target this library will be an executable.

Leaving the various framework and compatibility arguments for the moment, what negative impact do you see from adding netstandard2.1 as an additional target?

At the risk of sounding like a broken record, it's a single-line change without any other source changes. It needs no maintenance, and no additional components for developers. It widens the market availability for the library significantly and imposes no additional work on the maintainers.

Where's the downside?

**Comment by Perksey** [comment]
Thanks for the response Quahu. It seems there's a lot of heat in this thread so we should probably tone it down. If you have made up your mind on this PR just closing it is probably the most productive way. Name calling, or arguing is not how I want to be spending my Saturday afternoon nor do I imagine any of you want to either.

This PR would be useful for Remora.Discord, as this library is currently in need of a commands library and Nihlus wanted to use Qmmands ~~make Qmmands the official one~~. However, .NET Standard 2.0 not being supported proves to be a blocker. It's perfectly fine, just means we have to go out of our way to make yet another library (which is not the desirable path) when there's a perfectly good one sitting right here (Ultz's Volte Discord bot uses it and it's been lovely with basically no issues, but figured Greem is better suited to weigh in as I'm not directly involved in Ultz's Discord-related projects and Greem would likely be responsible for maintaining a fork if we choose to make one)

---

EDIT: in response to Nihlus

**Comment by Nihlus** [comment]
Clarification; this has nothing to do with using Qmmands as the "official" (whatever that might mean) command library for Remora.Discord. I simply saw a library I liked the look of, wanted to use, and couldn't because of targeting problems.

**Comment by Quahu** [comment]
The reason I've dropped .NET Standard in the first place is exactly what this PR is - special backwards-compatible code that pollutes the codebase and has to be maintained throughout the future versions.

The primary use of this library is in Discord bots, so it's natural for them to constantly change as Discord itself does that (it just deprecated another API version), hence maintainers of the bots have to be used to keep up-to-date all the time and I don't consider updating a problem (if I drop an old .NET Core target in exchange for a newer one).
So, unless you're planning on running a Discord bot on the jokingly above-mentioned "smart fridges", i.e. anything that isn't a machine that should be used for hosting one, the lack of .NET Standard target shouldn't be an issue.

Please instead consider retargeting your library to .NET Core, as I genuinely see no reason not to do so.

**Comment by Nihlus** [comment]
And once again, I reiterate - if you target netstandard2.1 instead, you don't need the backwards-compatible code. It has the same API surface that you're using right now, and will allow a wider range of use cases until net5 is released and the ecosystem moves on.

**Comment by Perksey** [comment]
I do see what you're getting at and as library author that is your decision to make, it's just it blocks a wide range of downstream consumers that have chosen to support these targets. However if you're okay with that, just close this pull request and we can get on with our Saturdays :D

**Comment by Quahu** [comment]
October is truly spooky.

**Comment by Nihlus** [comment]
I'd also like to thank everyone for their professionalism, respect, and understanding throughout these discussions. Being called a whore and an autistic minion, as well as seeing speculations about my sexual orientation, really warms my heart. I do still hope you'll continue considering (and hopefully implementing) .NET Standard 2.1 as a target.

unknown (1)

**Comment by Quahu** [comment]
> I'd also like to thank everyone for their professionalism, respect, and understanding throughout these discussions.

As would I, so please do not bring unrelated parties from a Discord chat into a PR that I thought was supposed to be handled professionally. Additionally, check the same channel you linked for any additional explanation.

**Pull request closed without merging**